editor/plugins/color_channel_selector.cpp

```diff
@@ -0,0 +1,134 @@
+#include "scene/gui/button.h"
+#include "scene/resources/style_box_flat.h"
+
+ColorChannelSelector::ColorChannelSelector() :
+		buttons() {
+	HBoxContainer *container = memnew(HBoxContainer);
+	container->add_theme_constant_override("separation", 0);
+
+	create_button(0, "R", container);
+	create_button(1, "G", container);
+	create_button(2, "B", container);
+	create_button(3, "A", container);
+
+	// Use a bit of transparency to be less distracting.
+	set_modulate(Color(1, 1, 1, 0.6));
+
+	add_child(container);
+}
+
+void ColorChannelSelector::_notification(int p_what) {
+	if (p_what == NOTIFICATION_THEME_CHANGED) {
+		// It is a common pattern in the editor to use `NOTIFICATION_THEME_CHANGED` to setup theme overrides that need
+		// access to the editor's theme. Unfortunately, in our case, this leads to infinite recursion.
+		if (_setting_up_theme) {
+			return;
+		}
+		_setting_up_theme = true;
+		// PanelContainer's background is invisible in the editor. We need a background.
+		// And we need this in turn because buttons don't look good without background (for example, hover is transparent).
+		Ref<StyleBox> bg_style = get_theme_stylebox(SceneStringName(panel), "TabContainer");
+		ERR_FAIL_COND(bg_style.is_null());
+		bg_style = bg_style->duplicate();
+		// The default content margin makes the widget become a bit too large. It should be like mini-toolbar.
+		const float editor_scale = EditorScale::get_scale();
+		bg_style->set_content_margin(SIDE_LEFT, 1.f * editor_scale);
+		bg_style->set_content_margin(SIDE_RIGHT, 1.f * editor_scale);
+		bg_style->set_content_margin(SIDE_TOP, 1.f * editor_scale);
+		bg_style->set_content_margin(SIDE_BOTTOM, 1.f * editor_scale);
+		add_theme_style_override(SceneStringName(panel), bg_style);
+		_setting_up_theme = false;
+	}
+}
+
+void ColorChannelSelector::set_available_channels_mask(uint32_t p_mask) {
+	for (unsigned int i = 0; i < CHANNEL_COUNT; ++i) {
+		const bool available = (p_mask & (1 << i)) != 0;
+		Button *button = buttons[i];
+		button->set_visible(available);
+	}
+}
+
+void ColorChannelSelector::on_button_toggled(bool p_unused_pressed) {
+	emit_signal("selected_channels_changed");
+}
+
+uint32_t ColorChannelSelector::get_selected_channels_mask() const {
+	uint32_t mask = 0;
+	for (unsigned int i = 0; i < CHANNEL_COUNT; ++i) {
+		Button *button = buttons[i];
+		if (button->is_visible() && buttons[i]->is_pressed()) {
+			mask |= (1 << i);
+		}
+	}
+	return mask;
+}
+
+// Helper
+Vector4 ColorChannelSelector::get_selected_channel_factors() const {
+	Vector4 channel_factors;
+	const uint32_t mask = get_selected_channels_mask();
+	for (unsigned int i = 0; i < 4; ++i) {
+		if ((mask & (1 << i)) != 0) {
+			channel_factors[i] = 1;
+		}
+	}
+	return channel_factors;
+}
+
+void ColorChannelSelector::create_button(unsigned int p_channel_index, const String &p_text, Control *p_parent) {
+	CRASH_COND(p_channel_index >= CHANNEL_COUNT);
+	CRASH_COND(buttons[p_channel_index] != nullptr);
```

Comment:
`CRASH_COND` should be used extremely sparingly. Making the engine crash for end users because developers messed up is not the Godot philosophy.

These should be `ERR_FAIL_COND` most likely.

Reply:
I used `CRASH_COND` here because this function is not supposed to even be called in any other way than in the constructor (it was for code re-use), and it's extremely obvious at development time if that would fail. There is no user code path that would cause it. I could as well remove it by now since I tested the code quite a few times, I only left it to keep the function obvious.
I can replace it to be totally safe, or wrap it in `DEV_ENABLED`.

Reply:
I'd say `ERR_FAIL_COND` should still be fine for this, especially if you don't foresee a risk that the contract would easily be broken. Alternatively `DEV_ASSERT`, but it still feels a bit heavy handed to me.